### PR TITLE
[air/output] Restructure AIR reporter classes

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

We currently have two different classes for controlling outputs: Callbacks and Reporters. This division is confusing in development. Instead, we should double down on the reporter class and only use the callback to forward formatting requests to the Controller. This is also required to enable the (experiment-level) reporter reacting to trial-level events (such as results).

This PR consolidates the reporter callbacks and simplifies output configuration.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/tests/output/test_output.py [hunk 1]
```diff
@@ -81,16 +81,16 @@ def test_get_time_str():
 
 
 def test_get_trials_by_state():
-    t1 = Trial("__fake")
+    t1 = Trial("__fake", stub=True)
```
Comment (krfricke) [inline python/ray/tune/tests/output/test_output.py:84]: These changes are unrelated but it speeds up the test suite to run in <100ms.

File: python/ray/tune/tune.py [hunk 2]
```diff
@@ -1066,14 +1066,21 @@ class and registered trainables.
...
             air_verbosity,
             num_samples=search_alg.total_samples,
             entrypoint=_entrypoint,
             metric=metric,
             mode=mode,
             config=config,
+            progress_metrics=progress_metrics,
         )
+        for callback in callbacks:
+            if isinstance(callback, AirResultProgressCallback):
```
Comment (woshiyyya) [inline python/ray/tune/tune.py:1081]: Can we directly create the `AirResultProgressCallback` with progress reporter in `_create_default_callbacks`?

File: python/ray/tune/experimental/output.py [hunk 3]
```diff
@@ -621,13 +647,109 @@ def _time_heartbeat_str(self):
...
+        iteration: int,
+        trials: List[Trial],
+        trial: Trial,
+        checkpoint: "_TrackedCheckpoint",
+        **info,
+    ):
+        if self._verbosity < self._intermediate_result_verbosity:
+            return
+        # don't think this is supposed to happen but just to be save.
+        saved_iter = "?"
```
Comment (woshiyyya) [inline python/ray/tune/experimental/output.py:726]: Can we set `saved_iter = 0` by default?

File: python/ray/tune/experimental/output.py [hunk 4]
```diff
@@ -578,8 +591,15 @@ class ProgressReporter:
     _heartbeat_freq = 30  # every 30 sec
     # to be updated by subclasses.
     _heartbeat_threshold = None
+    _start_end_verbosity = None
+    _intermediate_result_verbosity = None
```
Comment (woshiyyya) [inline python/ray/tune/experimental/output.py:595]: Can the users specify these two verbosity independently, Or it's just for internal use?

**Review by woshiyyya — COMMENTED** [review]
Thanks Kai for refractor the progress reporter. Generally looks good to me. Left some comments.

One question: Is it possible to make `ProgressReporter` a subclass of `Callback`? Seems that `AirResultProgressCallback` is just a thin wrapper over the ProgressReporter.

File: python/ray/tune/experimental/output.py (hunk 3 above)
Comment (krfricke) [reply python/ray/tune/experimental/output.py:726]: Ah this is just moved from below - I'll leave it as is

File: python/ray/tune/experimental/output.py (hunk 4 above)
Comment (krfricke) [reply python/ray/tune/experimental/output.py:595]: We haven't really worked on the configuration interface, yet, but these are just internal flags and shouldn't be specified.

**Comment by krfricke** [comment]
> Thanks Kai for refractor the progress reporter. Generally looks good to me. Left some comments.
> 
> One question: Is it possible to make `ProgressReporter` a subclass of `Callback`? Seems that `AirResultProgressCallback` is just a thin wrapper over the ProgressReporter.

Ah yeah that actually makes sense, thanks! Updated the PR

File: python/ray/tune/experimental/output.py (hunk 4 above)
Comment (woshiyyya) [reply python/ray/tune/experimental/output.py:594]: got it!

File: python/ray/tune/tune.py (hunk 2 above)
Comment (woshiyyya) [reply python/ray/tune/tune.py]: I see. That makes sense. The apis will be simplified after we have a context singleton.

**Review by woshiyyya — APPROVED** [review]

**Pull request merged**